rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md

```diff
@@ -0,0 +1,581 @@
+refresh either. Even though the data refresh process takes considerable time
+before it's ready to delete the previous index, we don't want to play with race
+conditions.
+
+Rather than trying to use a lock, we will read the DAG status directly from the
+DAG and fail the filtered index creation DAG run if the corresponding data
+refresh DAG is running, e.g.:
+
+```py
+from data_refresh.dag_factory import image_data_refresh, audio_data_refresh
+
+def filtered_index_creation_dag():
+    # ...
+    if model_name == "image":
+        if image_data_refresh.get_active_runs():
+            # fail dag
+    # etc for other media types
+```
+
+In the data refresh, however, rather than failing if the filtered index creation
+DAG is running, we'll merely tell it to wait, using the
+[`ExternalTaskSensor`](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/sensors/external_task/index.html#airflow.sensors.external_task.ExternalTaskSensor).
+That way it'll already be running and no new filtered index creation runs can be
+triggered between some potential lag time of the first one finishing before the
+data refresh is ready.
+
+When the data refresh DAG triggers filtered index creation, it should pass the
+suffix of the newly created origin index so that they match. We can identify
+filtered indexes created during data refresh because they will have a matching
+suffix to the origin index. Manual runs of the filtered index creation DAG
+should exclude this so that a new suffix is created to prevent clashing with the
+existing filtered index that was created for the previous data refresh run.
+
+The new DAG should be modelled on the existing data refresh DAG's patterns for
+making requests to the ingestion server. It must do the following:
+
+1. Read the existing filtered index alias destination and save it (for deleting
+   later)
+   - The
+     [data refresh DAG does this here](https://github.com/WordPress/openverse-catalog/blob/75990f356ed6d79601ecf3ecd96e48c3932acb47/openverse_catalog/dags/data_refresh/data_refresh_task_factory.py#L167-L174).
+1. Create a new suffix to use if one is not provided in the variables
+1. Call the `create_and_populate_filtered_index` action in the ingestion server
+   and wait for completion
+   - See existing data refresh DAG strategy for making and waiting for ingestion
+     server requests
+1. Call the `point_alias`[^no-promote-action] action in the ingestion server and
+   wait for completion.
+1. Delete the previous filtered index whose canonical name we retrieved in the
+   first step (`DELETE_INDEX` action)
+
+[^no-promote-action]:
+    We cannot use the `promote` action used by the regular data refresh flow
+    because while that action does point the alias to the newly created index,
+    it also promotes the API tables, which is not necessary for filtered index
+    creation. Therefore, we use the `point_alias` action which only points
+    aliases and nothing else.
```

Comment:
Forgive my pedantry here, but should the various actions all be capitalized? The mixed case is throwing me off in thinking that they might be referring to different things:
```suggestion
1. Call the `POINT_ALIAS`[^no-promote-action] action in the ingestion server and
   wait for completion.
1. Delete the previous filtered index whose canonical name we retrieved in the
   first step (`DELETE_INDEX` action)

[^no-promote-action]:
    We cannot use the `PROMOTE` action used by the regular data refresh flow
    because while that action does point the alias to the newly created index,
    it also promotes the API tables, which is not necessary for filtered index
    creation. Therefore, we use the `POINT_ALIAS` action which only points
    aliases and nothing else.
```

Reply:
For clarity yes. These actions correspond 1-to-1 with the `TableIndexer` methods, but the API does ask for them in caps.